zenodo/modules/deposit/views_rest.py

```diff
@@ -53,3 +69,176 @@ def language():
+    @wraps(f)
+    def inner(self, *args, **kwargs):
+        mimetype = request.headers.get('Content-Type')
+        if mimetype not in _mimetype_whitelist():
+            abort(400, '"{}" is not an acceptable mimetype.'.format(mimetype))
+        return f(self, *args, **kwargs)
+    return inner
+
+
+read_permission = check_oauth2_scope(
+    lambda record: deposit_read_permission_factory(record=record).can(),
+    write_scope.id, metadata_read_scope.id)
+
+update_permission = check_oauth2_scope(
+    lambda record: deposit_update_permission_factory(record=record).can(),
+    write_scope.id, metadata_write_scope.id)
+
+
+METAFILE_PERMISSIONS = {
+    'read': read_permission,
+    'update': update_permission,
+}
+
+
+def check_deposit_permission(action):
+    """Deposit permission decorator."""
+    def _permission_builder(f):
+        @wraps(f)
+        def decorator(self, record=None, *args, **kwargs):
+            if not METAFILE_PERMISSIONS[action](record=record).can():
+                abort(403)
+            return f(self, record=record, *args, **kwargs)
+        return decorator
+    return _permission_builder
+
+
+def get_or_create_metadata_bucket(record):
+    if not record['_buckets'].get('metadata'):
+        metadata_bucket = Bucket.create(
+            quota_size=current_app.config['ZENODO_BUCKET_QUOTA_SIZE'],#lignusfix maybe define a smaller limit
+            max_file_size=current_app.config['ZENODO_MAX_FILE_SIZE'],
+        )
+        # locked or unlocked?
+        record['_buckets']['metadata'] = str(metadata_bucket.id)
+        record.commit()
+        RecordsBuckets.create(record=record.model, bucket=metadata_bucket)
+    else:
+        metadata_bucket = Bucket.query.get(record['_buckets']['metadata'])
+    return metadata_bucket
+
+
+class DepositMetadataResource(MethodView):
+    """Deposit metadata resource."""
+
+    @validate_metadata_mimetype
+    @pass_record
+    #@check_deposit_permission('read')
+    def get(self, pid, record, **kwargs):
+        """Get metadata."""
+        mimetype = request.headers.get('Content-Type')
+        if mimetype in record.metafiles:
+            fileobj = record.metafiles[mimetype]
+            return fileobj.obj.send_file(trusted=True)  # XXX: DANGER-ZONE
+        return abort(
+            404, 'No metadata exists for the mimetype "{}".'.format(mimetype))
+
+    @validate_metadata_mimetype
+    @pass_record
+    #@check_deposit_permission('update')
+    def put(self, pid, record, **kwargs):
+        """Create or replace metadata."""
+        stream, content_length, content_md5 = \
+            current_files_rest.upload_factory()
+        mimetype = request.headers.get('Content-Type')
+        metadata_bucket = get_or_create_metadata_bucket(record)
+        if record.is_published(): # maybe pass record to deposit var to be easier to read
+            pub_recid, published_record = record.fetch_published()
+        
+        with db.session.begin_nested():
```

Comment:
This logic could be extracted into its own `link_record_metabucket(record, bucket)` function